File: 688-knight-probability-in-chessboard/knight-probability-in-chessboard.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int N;
    unordered_map<string,double> map;
    vector<pair<int,int>> dirn={{-2,-1},{-1,-2},{1,-2},{2,-1},{2,1},{1,2},{-1,2},{-2,1}};
    double solve(int r,int c,int k){
        if(r<0||r>=N||c<0||c>=N){ //out of bounds
            return 0;
        }
        if(k==0){//0 steps
            return 1;
        }
        string key=to_string(k)+"_"+to_string(r)+"_"+to_string(c);
        if(map.find(key)!=map.end()){
            return map[key];
        }

        double ans=0;
        for(auto &dir:dirn){
            int nrow=r+dir.first;
            int ncol=c+dir.second;
            ans+=(double)solve(nrow,ncol,k-1);
        }
        return map[key]=(double)ans/(8.0);
    }
    double knightProbability(int n, int k, int row, int column) {
        N=n;
        return solve(row,column,k);
    }
};

File: 688-knight-probability-in-chessboard/knight-probability-in-chessboard.java
import java.util.HashMap;

class Solution {
    private int[][] dirn={{-2,-1},{-1,-2},{1,-2},{2,-1},{2,1},{1,2},{-1,2},{-2,1}};
    public double knightProbability(int n, int k, int r, int c) {
        HashMap<String,Double> hm=new HashMap<>();
        return find(n,k,r,c,hm);
    }
    public double find(int n, int k, int r, int c,HashMap<String,Double> hm) {
        if(isvalid(r,c,n)==false) return 0;
        if(k==0) return 1;
        String key=r+","+c+","+k;
        if(hm.containsKey(key)) return hm.get(key);
        double prob=0;
        for(int i=0;i<dirn.length;i++){
            prob+=find(n,k-1,r+dirn[i][0],c+dirn[i][1],hm)/8.0;
        }
        hm.put(key,prob);
        return prob;
    }
    private boolean isvalid(int r, int c,int n){
        if(r<0||r>=n||c<0||c>=n) return false;
        return true;
    }
}

//tle
// class Solution {
//     private int[][] dirn={{-2,-1},{-1,-2},{1,-2},{2,-1},{2,1},{1,2},{-1,2},{-2,1}};
//     public double knightProbability(int n, int k, int r, int c) {
//         return find(n,k,r,c);
//     }
//     public double find(int n, int k, int r, int c) {
//         if(isvalid(r,c,n)==false) return 0;
//         if(k==0) return 1;
//         double prob=0;
//         for(int i=0;i<dirn.length;i++){
//             prob+=find(n,k-1,r+dirn[i][0],c+dirn[i][1])/8.0;
//         }
//         return prob;
//     }
//     private boolean isvalid(int r, int c,int n){
//         if(r<0||r>=n||c<0||c>=n) return false;
//         return true;
//     }
// }